### DISCUSSION TITLE
High amount of DB connections

### DISCUSSION THREAD

**Discussion by rbustos** [postalserver/postal#1443] [Help with using Postal]
Is there a technical reason to keep postal opening such amount of connections to the database?

We do run our MySQL in a cluster mode (Galera4 + MySQL8) and we see a lot of connections coming from postal instances (2 in our case). (we were using Galera3+MySQL5.7 as well)

It cause interest from our DBA perspective, since under intensive usage it really increase the amount of memory/cpu usage of whole cluster. So, wonder why it doesn't looks like other transactional app? where those tends to reuse one or more  threads in a keep alive fashion tunnel.

**Comment by willpower232** [comment]
Its likely that your Postal instances are seeing traffic even if you aren't using them, thanks to likely having an open SMTP port, so between that and all the separate processes Postal makes use of, I wouldn't be surprised at the volume of SQL usage. We aren't running a separate SQL setup (yet) so I don't have first hand experience of this.

That said, I'm not super familiar with ruby, with or without rails, so its possible there are config improvements available at some level.

**Comment by catphish** [comment]
How many connections are you seeing? There shouldn't be lots, and they should be reused :slightly_frowning_face:

**Comment by rbustos** [comment]
**DB & Workers:**
```
main_db:
  pool_size: 50
  reconnect: true

message_db:
  pool_size: 50
  reconnect: true

workers:
  quantity: 16
  threads: 20
```

**Per server:** 
```
root@postal-15:~# netstat -a| grep mysql| wc -l
466
root@postal-15:~# netstat -a| grep mysql | grep TIME_WAIT | wc -l
9
root@postal-15:~# netstat -a| grep mysql | grep ESTABLISHED | wc -l
453
```
**Per app server:**
```
root@postal-15:~# netstat -ap| grep mysql| grep fas| wc -l
34
root@postal-15:~# netstat -ap| grep mysql| grep wor| wc -l
415
root@postal-15:~# netstat -ap| grep mysql| grep smt| wc -l
2
```

**Some screenshots about real DB usage by two postal instances under normal use (not running a large delivery/processing):**
![image](https://user-images.githubusercontent.com/263415/95994392-9d80f900-0e06-11eb-9038-9845672b1d4f.png)

![image](https://user-images.githubusercontent.com/263415/95995664-2fd5cc80-0e08-11eb-8371-e61d77ecc066.png)

**Comment by rbustos** [comment]
@catphish : do you have same behavior on your infra?
